lua/autorun/chatsound-minecraft-subtitles.lua

```diff
@@ -0,0 +1,202 @@
+			subtitles[id].alpha = 255
+			subtitles[id].count = ( subtitles[id].count or 0 ) + 1
+			subtitles[id].name = subtitles[id].count > 1
+				and '%s (x%i)' % { id, subtitles[id].count }
+				 or id
+		end,
+		HUDPaint = function( self )
+			local now = CurTime()
+			local x, y = ScrW() - self.screenPaddingInt, math.Round( ScrH() * ( 1 - self.screenPaddingPercent ) ) - self.screenPaddingInt
+			local widestName = 0
+
+			self:setMinecraftFont()
+
+			for k, s in pairs( self.subtitles ) do
+				if s.deadline < now then
+					if s.deadline + self.fadeOutDuration < now then
+						self.subtitles[k] = nil
+						continue
+					end
+
+					self.subtitles[k].alpha = ( 1 - ( now - s.deadline ) / self.fadeOutDuration ) * 255
+				end
+
+				s.width = surface.GetTextSize( s.name )
+				widestName = math.max( widestName, s.width )
+			end
+
+			local w = self.minStaticWidth + widestName
+			local h = table.Count( self.subtitles ) * self.fontHeight
+			x = x - w
+			y = y - h
+
+			surface.SetDrawColor( self.backgroundColor.r, self.backgroundColor.g, self.backgroundColor.b, self.backgroundColor.a )
+			surface.DrawRect( x, y, w, h )
+
+			local i = 0
+			local myPos = LocalPlayer():GetShootPos()
+			local yaw = LocalPlayer():EyeAngles().yaw
+
+			for k, s in pairs( self.subtitles ) do
+				local direction = s.position - LocalPlayer():GetShootPos()
+				local angle = 180
+
+				if direction:Length2DSqr() > self.minDistance2 then
+					angle = math.Round( yaw - direction:Angle().yaw ) % 360
+
+					if angle >= 180 then
+						angle = angle - 360
+					end
+				end
+
+				-- if angle < -self.minAngle or angle > ( 180 - self.minAngle ) then
+				if angle == 180 or angle < -self.minAngle then
+					self:drawMinecraftText( self.leftArrow, x + self.padding, y + i * self.fontHeight, s.alpha )
+				end
+
+				-- if angle > self.minAngle or angle < -( 180 - self.minAngle ) then
+				if angle == 180 or angle > self.minAngle then
+					self:drawMinecraftText( self.rightArrow, x + w - self.padding - self.rightArrowWidth, y + i * self.fontHeight, s.alpha )
+				end
+
+				local center = ( widestName - s.width ) / 2
+				self:drawMinecraftText( s.name, x + self.padding + self.leftArrowWidth + self.gap + center, y + i * self.fontHeight, s.alpha )
+
+				i = i + 1
+			end
+		end,
+	}
+}
+
+function ChatsoundMinecraftSubtitles:init()
+	self.minDistance2 = self.minDistance * self.minDistance
+
+	self:calcScales()
+	self:toggle()
+
+	cvars.AddChangeCallback( 'chatsounds_minecraft_subtitles_enable', function( _, oldValue, value )
+		self:toggle()
+	end, 'chatsounds_minecraft_subtitles_enable' )
+
+	cvars.AddChangeCallback( 'chatsounds_minecraft_subtitles_size', function( _, oldValue, value )
+		self:calcScales()
+	end, 'chatsounds_minecraft_subtitles_size' )
+end
+
+function ChatsoundMinecraftSubtitles:calcScales()
+	local size = sizeCV:GetInt()
+	local fontSize = self.fontScale * size
+	self.shadowOffset = self.shadowOffsetScale * size
+
+	surface.CreateFont( self.fontName, {
```

Comment:
creating the same font multiple times like this isn't a good idea because fonts aren't dynamic and they're really taxing on the game

Reply:
Do I use different name for each size then?

Reply:
Changing font size will still create multiple fonts, but at the launch time only one font will be created.

Reply:
this game sucks so "ideally" you would cope with it and only create one font
if you feel that it's needed make 2-3 fonts for scale options and create them at the top yeah

Reply:
I had all 8 sizes being created on init call and didn't have any problems. What's the problem with creating font on size change?

Reply:
they are not meant for that which leads to instability
there's a warning on the wiki page for createfont because they're static and expensive

Reply:
What about this?

Reply:
https://github.com/Metastruct/incubator/pull/10/commits/9e772a4ca2aaa3a51ca4356b63c5ce4e5a6061da